administrator/components/com_users/models/user.php

```diff
@@ -189,6 +190,65 @@ public function save($data)
 		$pk   = (!empty($data['id'])) ? $data['id'] : (int) $this->getState('user.id');
 		$user = JUser::getInstance($pk);
 
+		// Make sure the activation / reset token is invalidated when the account mail is changed and a token exists.
+		if ($user->email != $data['email'] && $pk != 0 && !empty($user->activation) && !$user->block)
+		{
+			$app = JFactory::getApplication();
+
+			// Compile the notification mail values.
+			$data['fromname']   = $app->get('fromname');
+			$data['mailfrom']   = $app->get('mailfrom');
+			$data['sitename']   = $app->get('sitename');
+			$data['siteurl']    = JUri::root();
+			$data['activation'] = JApplicationHelper::getHash(JUserHelper::genRandomPassword());
+
+			// Set the link to activate the user account.
+			$linkMode = (int) $app->get('force_ssl', 0) == 2 ? Route::TLS_FORCE : Route::TLS_IGNORE;
+			$data['activate'] = Route::link('site', 'index.php?option=com_users&task=registration.activate&token=' . $data['activation'], false, $linkMode);
```

Comment:
```suggestion
			$data['activate'] = Route::link('site', 'index.php?option=com_users&task=registration.activate&token=' . $data['activation'], false, $linkMode, true);
```
This addresses the problem found by @ReLater